cpp/docs/TRANSITIONGUIDE.md

```diff
@@ -0,0 +1,206 @@
+column some_function( cudf::table const& input_table, cudf::table * input_output_table, int size){
+  column col;
+  // Do stuff to create output col
+  ...
+  // Return the result column of this function
+  return col;
+}
+}
+```
+
+And the usage would be:
+```
+cudf::table input;
+cudf::table in_out;
+int size;
+column output = cudf::some_function(input, &in_out, size);
+```
+
+### Multiple Return Values
+
+Sometimes it is necessary for functions to have multiple outputs. There are a few ways this can be done in C++ (including creating a `struct` for the output). One convenient way to do this is using `std::tie` and `std::make_tuple`.
+
+```
+auto return_multiple_outputs(void){
+  cudf::table out0;
+  cudf::table out1;
+  ...
+  // Do stuff with out0, out1
+  
+  // Return a tuple of all the outputs
+  return std::make_tuple(out0, out1);
+}
+
+
+cudf::table out0;
+cudf::table out1;
+std::tie(out0, out1) = cudf::return_multiple_outputs();
+```
+
+
+## Error Checking
+
+The `gdf_error` codes are deprecated. Instead, `libcudf` functions should prefer throwing exceptions.
+
+In place of the `GDF_REQUIRES` convenience macro, the `CUDF_EXPECTS` macro is provided in `cudf/cpp/src/utilities/error_utils.hpp`. 
+
+Similar to `GDF_REQUIRES`, `CUDF_EXPECTS` will throw an exception if a condition is violated. The first argument is the conditional expression that can be evaluated as a `boolean` expression and is expected to resolve to `true` under normal conditions. If the conditional evaluates to `false`, then an error has occurred and an instance of `cudf::logic_error` is thrown. The second argument to `CUDF_EXPECTS` is a short description of the error that has occurred if the conditional is false. 
+
+Example usage: 
+
+```
+CUDF_EXPECTS(lhs->dtype == rhs->dtype, "Column type mismatch");
+```
+
+There are times where a particular code path, if reached, should indicate an error no matter what. For example, in the `default` case of a `switch` statement where the only valid code paths are in one of the `case` statements. 
+
+For these cases, the `CUDF_FAIL` convenience macro should be used. This is effectively the same as doing `CUDF_EXPECTS(false, reason)`. 
+
+Example:
+
+```
+CUDF_FAIL("This code path should not be reached.");
+```
+
+### CUDA Error Checking
+
+Checking for the succesful completion of CUDA runtime API functions should continue to be done via the `CUDA_TRY` macro. Note that this macro has been updated to throw an exception of type `cudf::cuda_error` if the return value of the CUDA API does not return `cudaSuccess`. The thrown exception will include a description of the CUDA error code that occurred in it's `what()` message.
+
+Example:
+
+```
+CUDA_TRY( cudaMemcpy(&dst, &src, num_bytes) );
+```
+
+## Testing
+
+### ```column_wrapper```
+
+The `column_wrapper<T>` class template is defined in `cudf/cpp/tests/utilities/column_wrapper.cuh` to simplify the creation and management of `gdf_column`s for the purposes of unit testing. 
```

Comment:
It doesn't simplify things all that much, and needs some work... perhaps suggest someone work on improving it and link to issue #1178.

... need to end the review here as I have to leave.